heka/io_modules/decoders/heka/json.lua

```diff
@@ -34,18 +48,38 @@ Decode and inject the resulting message
 --]]
 
 -- Imports
+local module_name   = ...
+local module_cfg    = require "string".gsub(module_name, "%.", "_")
 local cjson = require "cjson"
 
 local pairs = pairs
 local type  = type
 
 local inject_message = inject_message
 
+local cfg = read_config(module_cfg) or {}
+assert(type(cfg) == "table", module_cfg .. " must be a table")
+cfg.flatten_delimiter = cfg.flatten_delimiter or "."
+assert(type(cfg.flatten_delimiter) == "string", module_cfg .. ".flatten_delimiter must be a string")
+fields_prefix = "Fields" .. cfg.flatten_delimiter
+
 local M = {}
 setfenv(1, M) -- Remove external access to contain everything in the module
 
 function decode(data, dh)
     local msg = cjson.decode(data)
+    if cfg.preserve_metadata then
+        if type(msg.Fields) == "table" then
+            for k,v in pairs(msg.Fields.Fields) do
```

Comment:
I need to see your original input as this implies it does not conform to the Heka message schema as there is no nested Fields of Fields hash.